ops/testing.py

```diff
@@ -508,8 +508,13 @@ def attach_storage(self, storage_id: str) -> None:
             storage_id: The full storage ID of the storage unit being attached, including the
                 storage key, e.g. my-storage/0.
         """
-        if self._backend._storage_attach(storage_id) and self._hooks_enabled:
+        if self._backend._storage_attach(storage_id) and self.charm and self._hooks_enabled:
```

Comment:
what is the scenario in which one should be calling Harness.detach_storage before begin()?

Reply:
I could imagine some sloppy test setup code that attaches all storages, then goes through and unattaches a particular one before begin.  But in general, it doesn't seems like it should be necessary.

Reply:
If you meant "attach", then if you don't care about storage-attached events, but you want to do some testing that interacts with storage, then it can be useful to attach before  begin is called - avoiding the unwanted hook call.  pre-begin detach is less necessary and has always raised an error in that case.